File: monitor.py
import requests
import json
import time
import os
from datetime import datetime

TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN")
TELEGRAM_CHAT_ID = os.getenv("TELEGRAM_CHAT_ID")

CREATORS = {
    "Square-Creator-1df1e693e2192": "Creator 1",
    "Acqua_DY": "Acqua_DY",
    "Square-Creator-5dd415213": "Creator 3",
    "xiaoxiong": "xiaoxiong",
    "sanmageshuai": "sanmageshuai",
    "Square-Creator-4d698fecefd05": "Creator 6",
    "susea": "susea",
    "Square-Creator-19579394c90dc": "Creator 8",
    "Chungorcrypto": "Chungorcrypto",
    "SaMnAtIoN": "My Account (Test)",
}

POLL_INTERVAL = 45
SEEN_FILE = "seen.json"

def load_seen():
    if os.path.exists(SEEN_FILE):
        with open(SEEN_FILE, "r") as f:
            return json.load(f)
    return {}

def save_seen(seen):
    with open(SEEN_FILE, "w") as f:
        json.dump(seen, f)

def send_telegram(text):
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage"
    payload = {
        "chat_id": TELEGRAM_CHAT_ID,
        "text": text,
        "disable_web_page_preview": False
    }
    try:
        r = requests.post(url, json=payload, timeout=15)
        return r.status_code == 200
    except Exception as e:
        print("Telegram error:", e)
        return False

def fetch_posts(square_uid):
    url = "https://www.binance.com/bapi/composite/v2/friendly/pgc/content/queryUserProfilePageContentsWithFilter"
    params = {
        "targetSquareUid": square_uid,
        "timeOffset": -1,
        "filterType": "ALL"
    }
    headers = {
        "User-Agent": "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Mobile Safari/537.36",
        "Accept": "application/json",
        "clienttype": "android",
        "lang": "en"
    }
    try:
        r = requests.get(url, params=params, headers=headers, timeout=15)
        return r.json()
    except Exception as e:
        print(f"Fetch error for {square_uid}:", e)
        return None

def main():
    print("Starting Binance Square → Telegram monitor...")
    print(f"Polling every {POLL_INTERVAL} seconds")
    
    if not TELEGRAM_BOT_TOKEN or not TELEGRAM_CHAT_ID:
        print("ERROR: TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID not set")
        return

    seen = load_seen()

    while True:
        for uid, name in CREATORS.items():
            print(f"[{datetime.now().strftime('%H:%M:%S')}] Checking {name}...")
            data = fetch_posts(uid)

            if not data or data.get("code") != "000000":
                print("  → No data or error")
                continue

            posts = data.get("data", {}).get("list") or data.get("data", {}).get("vos") or []
            
            if not posts:
                print("  → Empty list")
                continue

            for post in posts[:5]:
                post_id = str(post.get("id") or post.get("contentId") or "")
                if not post_id:
                    continue

                key = f"{uid}_{post_id}"
                if key in seen:
                    continue

                content = post.get("bodyTextOnly") or post.get("content") or post.get("title") or "No text"
                link = f"https://www.binance.com/en/square/post/{post_id}"

                message = f"📢 New post from {name}\n\n{content[:800]}\n\n🔗 {link}"
                
                if send_telegram(message):
                    print(f"  → Sent new post {post_id}")
                    seen[key] = int(time.time())
                    save_seen(seen)
                else:
                    print("  → Failed to send to Telegram")

        print(f"Sleeping {POLL_INTERVAL}s...\n")
        time.sleep(POLL_INTERVAL)

if __name__ == "__main__":
    main()

File: monitor.js
const { chromium } = require("playwright");
const http = require("http");

const PORT = process.env.PORT || 10000;
const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const CHAT_ID = process.env.TELEGRAM_CHAT_ID;

const CREATORS = [
  "SaMnAtIoN00",
  "Square-Creator-1df1e693e2192",
  "Acqua_DY",
  "Square-Creator-5dd415213",
  "xiaoxiong",
  "sanmageshuai",
  "Square-Creator-4d698fecefd05",
  "susea",
  "Square-Creator-19579394c90dc",
  "Chungorcrypto"
];

const seenPosts = new Set();

console.log("==========================================");
console.log("🚀 BINANCE SQUARE → TELEGRAM");
console.log("==========================================");
console.log("👥 10 CREATOR MONITOR");
console.log("==========================================");

CREATORS.forEach((creator, i) => {
  console.log(`${i + 1}. ${creator}`);
});

console.log("==========================================");

// --------------------------------------------------
// HEALTH SERVER
// --------------------------------------------------

const server = http.createServer((req, res) => {
  res.writeHead(200, {
    "Content-Type": "text/plain"
  });

  res.end("Binance Square Telegram monitor is running");
});

server.listen(PORT, "0.0.0.0", () => {
  console.log(`🌐 Health server running on port ${PORT}`);
});

// --------------------------------------------------
// TELEGRAM
// --------------------------------------------------

async function sendTelegram(message) {
  if (!BOT_TOKEN || !CHAT_ID) {
    console.log("⚠️ Telegram variables are missing");
    return;
  }

  try {
    const url =
      `https://api.telegram.org/bot${BOT_TOKEN}/sendMessage`;

    const response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        chat_id: CHAT_ID,
        text: message,
        disable_web_page_preview: false
      })
    });

    if (response.ok) {
      console.log("✅ Telegram message sent");
    } else {
      const text = await response.text();
      console.log("❌ Telegram error:", response.status, text);
    }
  } catch (error) {
    console.log("❌ Telegram connection error:", error.message);
  }
}

// --------------------------------------------------
// CHECK WHETHER RESPONSE LOOKS LIKE SQUARE DATA
// --------------------------------------------------

function looksInteresting(url, data) {
  const lowerUrl = url.toLowerCase();

  const interestingUrlWords = [
    "square",
    "pgc",
    "feed",
    "article",
    "buzz",
    "creator",
    "content",
    "recommend",
    "timeline",
    "author"
  ];

  if (interestingUrlWords.some(word => lowerUrl.includes(word))) {
    return true;
  }

  if (!data) {
    return false;
  }

  const text = JSON.stringify(data).toLowerCase();

  const interestingDataWords = [
    "squareauthorid",
    "authorname",
    "webLink".toLowerCase(),
    "sharelink",
    "contenttype",
    "cardtype",
    "username",
    "postid",
    "feedid"
  ];

  return interestingDataWords.some(word => text.includes(word));
}

// --------------------------------------------------
// EXTRACT POSTS FROM UNKNOWN RESPONSE
// --------------------------------------------------

function findPosts(obj, found = []) {
  if (!obj || typeof obj !== "object") {
    return found;
  }

  if (Array.isArray(obj)) {
    for (const item of obj) {
      findPosts(item, found);
    }

    return found;
  }

  // A Binance Square post usually contains an id and author information
  if (
    obj.id &&
    (
      obj.authorName ||
      obj.username ||
      obj.squareAuthorId ||
      obj.contentType ||
      obj.cardType
    )
  ) {
    found.push(obj);
  }

  for (const key of Object.keys(obj)) {
    try {
      findPosts(obj[key], found);
    } catch (_) {}
  }

  return found;
}

// --------------------------------------------------
// PROCESS POSSIBLE POSTS
// --------------------------------------------------

async function processPosts(posts) {
  if (!posts || posts.length === 0) {
    return;
  }

  for (const post of posts) {
    const id = String(post.id || "");

    if (!id) {
      continue;
    }

    const author =
      post.authorName ||
      post.username ||
      post.author ||
      "";

    const content =
      post.title ||
      post.subTitle ||
      post.content ||
      post.body ||
      "";

    const webLink =
      post.webLink ||
      post.shareLink ||
      `https://www.binance.com/en/square/post/${id}`;

    // Only show posts that could potentially belong to our creators.
    const authorMatches = CREATORS.some(
      creator =>
        String(author).toLowerCase() ===
        String(creator).toLowerCase()
    );

    if (authorMatches) {
      console.log("");
      console.log("🎯 ========================================");
      console.log("🎯 TARGET CREATOR POST FOUND!");
      console.log("🎯 ========================================");
      console.log(`👤 Author: ${author}`);
      console.log(`🆔 ID: ${id}`);
      console.log(`📝 Content: ${String(content).slice(0, 500)}`);
      console.log(`🔗 Link: ${webLink}`);
      console.log("🎯 ========================================");
      console.log("");

      if (!seenPosts.has(id)) {
        seenPosts.add(id);

        const message =
          `📢 <b>NEW BINANCE SQUARE POST</b>\n\n` +
          `👤 <b>${escapeHtml(author)}</b>\n\n` +
          `${escapeHtml(String(content).slice(0, 3000))}\n\n` +
          `⚡ <a href="${webLink}">OPEN IN BINANCE</a>`;

        await sendTelegram(message);
      }
    }
  }
}

// --------------------------------------------------
// HTML ESCAPE
// --------------------------------------------------

function escapeHtml(text) {
  return String(text)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

// --------------------------------------------------
// MAIN BROWSER MONITOR
// --------------------------------------------------

async function startMonitor() {
  console.log("");
  console.log("==========================================");
  console.log("🔧 BINANCE SQUARE API DIAGNOSTIC");
  console.log("==========================================");
  console.log("");

  console.log("👥 Creators we need to identify:");

  CREATORS.forEach((creator, i) => {
    console.log(`${i + 1}. ${creator}`);
  });

  console.log("");

  await sendTelegram(
    "🚀 <b>Binance Square monitor started</b>\n\n" +
    "👥 Monitoring 10 creators\n" +
    "🔎 Searching Binance Square API responses..."
  );

  console.log("✅ Telegram test successful");
  console.log("");
  console.log("🌐 Starting Chromium...");

  const browser = await chromium.launch({
    headless: true,
    args: [
      "--no-sandbox",
      "--disable-setuid-sandbox",
      "--disable-dev-shm-usage",
      "--disable-gpu"
    ]
  });

  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (X11; Linux x86_64) " +
      "AppleWebKit/537.36 (KHTML, like Gecko) " +
      "Chrome/124.0.0.0 Safari/537.36",
    viewport: {
      width: 1440,
      height: 900
    }
  });

  const page = await context.newPage();

  // ------------------------------------------------
  // REQUEST LOGGING
  // ------------------------------------------------

  page.on("request", request => {
    const url = request.url();

    if (
      url.includes("binance.com") &&
      (
        url.includes("/bapi/") ||
        url.includes("/api/") ||
        url.includes("square") ||
        url.includes("pgc") ||
        url.includes("feed")
      )
    ) {
      console.log(`➡️ API REQUEST: ${request.method()} ${url}`);
    }
  });

  // ------------------------------------------------
  // RESPONSE CAPTURE
  // ------------------------------------------------

  page.on("response", async response => {
    const url = response.url();

    if (!url.includes("binance.com")) {
      return;
    }

    const status = response.status();

    // We only care about likely API responses.
    const looksLikeApi =
      url.includes("/bapi/") ||
      url.includes("/api/") ||
      url.includes("square") ||
      url.includes("pgc") ||
      url.includes("feed") ||
      url.includes("buzz") ||
      url.includes("article");

    if (!looksLikeApi) {
      return;
    }

    console.log("");
    console.log("📡 RESPONSE");
    console.log(`STATUS: ${status}`);
    console.log(`URL: ${url}`);

    if (status !== 200) {
      console.log("⚠️ Non-200 response");
      return;
    }

    try {
      const contentType =
        response.headers()["content-type"] || "";

      if (!contentType.includes("json")) {
        console.log(`ℹ️ Content-Type: ${contentType}`);
        return;
      }

      const data = await response.json();

      if (looksInteresting(url, data)) {
        console.log("🔥 INTERESTING BINANCE RESPONSE");

        // Print only a manageable amount to Render logs.
        let preview = JSON.stringify(data);

        if (preview.length > 5000) {
          preview = preview.slice(0, 5000) +
            "... [TRUNCATED]";
        }

        console.log(preview);

        const posts = findPosts(data);

        if (posts.length > 0) {
          console.log(
            `🔎 Possible Square posts found: ${posts.length}`
          );

          await processPosts(posts);
        }
      }
    } catch (error) {
      console.log(
        `⚠️ Could not read JSON response: ${error.message}`
      );
    }
  });

  // ------------------------------------------------
  // CONSOLE LOGS FROM BINANCE PAGE
  // ------------------------------------------------

  page.on("console", msg => {
    const text = msg.text();

    if (
      text.includes("error") ||
      text.includes("Error") ||
      text.includes("square")
    ) {
      console.log(`🌐 PAGE: ${text.slice(0, 1000)}`);
    }
  });

  // ------------------------------------------------
  // PAGE ERRORS
  // ------------------------------------------------

  page.on("pageerror", error => {
    console.log(`⚠️ PAGE ERROR: ${error.message}`);
  });

  // ------------------------------------------------
  // OPEN BINANCE SQUARE
  // ------------------------------------------------

  console.log("");
  console.log("==========================================");
  console.log("🌐 OPENING BINANCE SQUARE");
  console.log("==========================================");

  try {
    await page.goto(
      "https://www.binance.com/en/square",
      {
        waitUntil: "domcontentloaded",
        timeout: 60000
      }
    );

    console.log("✅ Binance Square page loaded");
  } catch (error) {
    console.log(
      `⚠️ Page load warning: ${error.message}`
    );
  }

  // Give Binance time to make API requests.
  console.log("");
  console.log("⏳ Waiting for Binance API responses...");

  await page.waitForTimeout(30000);

  // ------------------------------------------------
  // CONTINUOUS MONITOR
  // ------------------------------------------------

  while (true) {
    try {
      console.log("");
      console.log("🔄 Reloading Binance Square...");

      await page.reload({
        waitUntil: "domcontentloaded",
        timeout: 60000
      });

      console.log("⏳ Waiting for Binance API responses...");

      await page.waitForTimeout(30000);
    } catch (error) {
      console.log(
        `⚠️ Reload error: ${error.message}`
      );

      await new Promise(resolve =>
        setTimeout(resolve, 10000)
      );
    }
  }
}

// --------------------------------------------------
// START
// --------------------------------------------------

startMonitor().catch(async error => {
  console.log("");
  console.log("❌ FATAL ERROR");
  console.log(error);

  try {
    await sendTelegram(
      "❌ <b>Binance Square monitor crashed</b>\n\n" +
      escapeHtml(error.message)
    );
  } catch (_) {}

  // Keep Render alive briefly so logs can be read.
  setTimeout(() => process.exit(1), 3000);
});
